service/storage_service.cc

```diff
@@ -5678,6 +5678,28 @@ future<raft_topology_cmd_result> storage_service::raft_topology_cmd_handler(shar
 
         switch (cmd.cmd) {
             case raft_topology_cmd::command::barrier:
+                // This barrier might have been issued by the topology coordinator
+                // as a step in enabling a feature, i.e. it noticed that all
+                // nodes support some feature, then issue the barrier to make
+                // sure that all nodes observed this fact in their local state
+                // (a node cannot revoke support for a feature after that), and
+                // after receiving a confirmation from all nodes it will mark
+                // the feature as enabled.
+                //
+                // However, it might happen that the node handles this request
+                // early in the boot process, before it did the second feature
+                // check that happens when the node updates its metadata
+                // in `system.topology`. The node might have committed a command
+                // that advertises support for a feature as the last node
+                // to do so, crashed and now it doesn't support it. This should
+                // be rare, but it can happen and we can detect it right here.
+                if (!_topology_updated_with_local_metadata) {
+                    // TODO: Somehow make it stop the boot process if it fails
+                    const auto& enabled_features = _topology_state_machine._topology.features.enabled_features;
+                    const auto unsafe_to_disable_features = _topology_state_machine._topology.features.calculate_not_yet_enabled_features();
+                    _feature_service.check_features(enabled_features, unsafe_to_disable_features);
```

Comment:
@gleb-cloudius in our discussion under the previous PR (https://github.com/scylladb/scylladb/pull/14722#discussion_r1285522121) you wrote that the feature check that happens in the `barrier` command handler should cause the node to shut down, however I have no idea how to propagate the failure from here in a sensible way to cause a shutdown. Currently, the startup procedure will eventually reach `check_features` in `update_topology_with_local_metadata` and this will shut down the node.

I'll be happy to implement it if you have a good suggestion on how to do it.

Reply:
We can print an appropriate error message an call ss::drain().